File: codes/javascript/claude/minimal/task_2/Solution.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

function reallocateBed(patient, hospitals) {
  const isHighRisk = patient.is_post_surgery;
  let result = {
    patient_id: patient.id,
    assigned_hospital: null,
    assigned_ward: null,
    assigned_bed: null,
    reallocated_patients: []
  };

  // Direct Assignment
  for (const hospital of hospitals) {
    if (isHighRisk && hospital.model !== 4) continue;
    if (!isHighRisk && hospital.model === 4) continue;

    for (const ward of hospital.wards) {
      if (isMatchingWard(patient, ward)) {
        for (const bed of ward.available_beds) {
          result.assigned_hospital = hospital.id;
          result.assigned_ward = ward.ward_name;
          result.assigned_bed = bed;
          return result;
        }
      }
    }
  }

  // Reallocation for high-risk patients
  if (isHighRisk) {
    for (const hospital of hospitals.filter(h => h.model === 4)) {
      for (const ward of hospital.wards) {
        if (isMatchingWard(patient, ward)) {
          for (const currentPatient of ward.current_patients) {
            if (currentPatient.days_in_hospital > 3 && !currentPatient.non_transferable) {
              const newHospital = findLowerLevelHospital(currentPatient, hospitals);
              if (newHospital) {
                result.assigned_hospital = hospital.id;
                result.assigned_ward = ward.ward_name;
                result.assigned_bed = currentPatient.bed_number;
                result.reallocated_patients.push({
                  patient_id: currentPatient.id,
                  new_hospital: newHospital.id,
                  new_ward: newHospital.ward.ward_name,
                  new_bed: newHospital.ward.available_beds[0]
                });
                return result;
              }
            }
          }
        }
      }
    }
  }

  return result;
}

function isMatchingWard(patient, ward) {
  return (
    (ward.gender_restriction === "No Restriction" || ward.gender_restriction === `${patient.gender} Only`) &&
    patient.age >= ward.age_restriction[0] &&
    patient.age <= ward.age_restriction[1] &&
    patient.special_requirements.every(req => ward.special_requirements.includes(req))
  );
}

function findLowerLevelHospital(patient, hospitals) {
  for (const hospital of hospitals.filter(h => h.model < 4)) {
    for (const ward of hospital.wards) {
      if (isMatchingWard(patient, ward) && ward.available_beds.length > 0) {
        return { id: hospital.id, ward: ward };
      }
    }
  }
  return null;
}

let inputData = '';
rl.on('line', (line) => {
  inputData += line;
}).on('close', () => {
  const { patient, hospitals } = JSON.parse(inputData);
  const result = reallocateBed(patient, hospitals);
  console.log(JSON.stringify(result));
});

File: codes/go/claude/minimal/task_2/Solution.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

func reallocateBed(patient map[string]interface{}, hospitals []interface{}) map[string]interface{} {
	result := map[string]interface{}{
		"patient_id":          nil,
		"assigned_hospital":   nil,
		"assigned_ward":       nil,
		"assigned_bed":        nil,
		"reallocated_patients": []interface{}{},
	}

	isHighRisk := patient["is_post_surgery"].(bool)

	// Direct Assignment
	for _, hospital := range hospitals {
		h := hospital.(map[string]interface{})
		if isHighRisk && h["model"].(float64) != 4 {
			continue
		}
		if !isHighRisk && h["model"].(float64) == 4 {
			continue
		}
		for _, ward := range h["wards"].([]interface{}) {
			w := ward.(map[string]interface{})
			if matchWardRequirements(patient, w) {
				availableBeds := w["available_beds"].([]interface{})
				if len(availableBeds) > 0 {
					result["patient_id"] = patient["id"]
					result["assigned_hospital"] = h["id"]
					result["assigned_ward"] = w["ward_name"]
					result["assigned_bed"] = availableBeds[0]
					return result
				}
			}
		}
	}

	// Reallocation for high-risk patients
	if isHighRisk {
		for _, hospital := range hospitals {
			h := hospital.(map[string]interface{})
			if h["model"].(float64) != 4 {
				continue
			}
			for _, ward := range h["wards"].([]interface{}) {
				w := ward.(map[string]interface{})
				if matchWardRequirements(patient, w) {
					currentPatients := w["current_patients"].([]interface{})
					for _, cp := range currentPatients {
						currentPatient := cp.(map[string]interface{})
						if currentPatient["days_in_hospital"].(float64) > 3 && !currentPatient["non_transferable"].(bool) {
							newHospital, newWard, newBed := findNewPlacement(currentPatient, hospitals)
							if newHospital != nil {
								result["patient_id"] = patient["id"]
								result["assigned_hospital"] = h["id"]
								result["assigned_ward"] = w["ward_name"]
								result["assigned_bed"] = currentPatient["bed_number"]
								result["reallocated_patients"] = []interface{}{
									map[string]interface{}{
										"patient_id":   currentPatient["id"],
										"new_hospital": newHospital["id"],
										"new_ward":     newWard["ward_name"],
										"new_bed":      newBed,
									},
								}
								return result
							}
						}
					}
				}
			}
		}
	}

	return result
}

func matchWardRequirements(patient map[string]interface{}, ward map[string]interface{}) bool {
	genderRestriction := ward["gender_restriction"].(string)
	if genderRestriction != "No Restriction" && genderRestriction != patient["gender"].(string)+" Only" {
		return false
	}

	ageRestriction := ward["age_restriction"].([]interface{})
	patientAge := patient["age"].(float64)
	if patientAge < ageRestriction[0].(float64) || patientAge > ageRestriction[1].(float64) {
		return false
	}

	patientRequirements := patient["special_requirements"].([]interface{})
	wardRequirements := ward["special_requirements"].([]interface{})
	for _, req := range patientRequirements {
		if !contains(wardRequirements, req) {
			return false
		}
	}

	return true
}

func findNewPlacement(patient map[string]interface{}, hospitals []interface{}) (map[string]interface{}, map[string]interface{}, interface{}) {
	for _, hospital := range hospitals {
		h := hospital.(map[string]interface{})
		if h["model"].(float64) == 4 {
			continue
		}
		for _, ward := range h["wards"].([]interface{}) {
			w := ward.(map[string]interface{})
			if matchWardRequirements(patient, w) {
				availableBeds := w["available_beds"].([]interface{})
				if len(availableBeds) > 0 {
					return h, w, availableBeds[0]
				}
			}
		}
	}
	return nil, nil, nil
}

func contains(slice []interface{}, item interface{}) bool {
	for _, v := range slice {
		if v == item {
			return true
		}
	}
	return false
}

func main() {
	var input struct {
		Patient   map[string]interface{} `json:"patient"`
		Hospitals []interface{}          `json:"hospitals"`
	}

	decoder := json.NewDecoder(os.Stdin)
	if err := decoder.Decode(&input); err != nil {
		fmt.Println("Error decoding input:", err)
		return
	}

	result := reallocateBed(input.Patient, input.Hospitals)

	encoder := json.NewEncoder(os.Stdout)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(result); err != nil {
		fmt.Println("Error encoding result:", err)
	}
}